README
Alien::Base is now distributed as part of Alien-Base.  Please see
https://github.com/Perl5-Alien/Alien-Build for details.

Alien::Base is a base class and framework for creating Alien 
distributions.  The goal of the project is to make things as simple and 
easy as possible for both developers and users of Alien modules.

Alien is a Perl namespace for defining dependencies in CPAN for 
libraries and tools which are not "native" to CPAN.  Alien modules will 
typically use the system libraries if they are available, or download 
the latest version from the internet and build them from source code. 
These libraries can then be used by other Perl modules, usually modules 
that are implemented with XS or FFI.

To install Alien::Base, use cpanminus:

 cpanm Alien::Base
 
If you have downloaded this from GitHub or locally, do

 cpanm --installdeps .
 cpanm .

Once installed you can read the project documentation using the perldoc 
command, or via the metacpan.org website.  If you are an Alien 
developer, see the authoring documentation:

 perldoc Alien::Base::ModuleBuild::Authoring
 perldoc Alien::Build::Manual::AlienAuthor

(this one installed in a separate module)

 http://metacpan.org/pod/Alien::Base::ModuleBuild::Authoring
 http://metacpan.org/pod/Alien::Build::Manual::AlienAuthor
 

The FAQ also contains hints on dealing with specific challenges, like 
dealing with specific tools:

 perldoc Alien::Base::FAQ
 http://metacpan.org/pod/Alien::Base::FAQ

If you are a user of an Alien module, you will likely be more interested 
in the Alien::Base interface itself:

 perldoc Alien::Base
 http://metacpan.org/pod/Alien::Base

For questions that aren't answered in the documentation, please feel 
free to contact the Alien::Base team:

 * IRC: #native on irc.perl.org
 * mailing list at https://groups.google.com/forum/#!forum/perl5-alien
 * GitHub issue tracker at https://github.com/Perl5-Alien/Alien-Base/issues




Changes
Revision history for Perl module Alien::Base.

0.043_01 2017-07-17 03:25:31 -0400
  - Migrate to Test2::V0
  - Documentation improvements (gh#187 jj++)

0.042 2017-05-23 14:40:02 -0400
  - Add diagnostics for when temp directory is not writable
    This manifests itself on Windows sometimes
  - Configure step will fail if temp directory is not writable

0.041 2017-04-02 18:00:52 -0400
  - Fix some minor metadata
  - Require Alien-Base-ModuleBuild 0.040 to ensure the separated
    version.

0.040 2017-03-29 01:30:46 -0400
  - Spin off Alien::Base::ModuleBuild into the separate distribution
    Alien-Base-ModuleBuild
  - Alien::Base::ModuleBuild is a prerequisite, but will be removed
    on October 1, 2017, so you need to update your Build.PL so that
    Alien::Base::ModuleBuild is a configure requires.

0.038 2017-03-06 00:16:18 -0500
  - Production release identical to 0.035_01 release

0.037_02 2017-03-05 03:20:00 -0500
  - Additional fixex for t/yy-system_installed.t

0.037_01 2017-03-03 05:29:54 +1100
  - cflags and cflags_static were flipped in 0.035_01 is now fixed
    (should fix gh#183)

0.036 2017-02-27 03:28:36 -0500
  - Production release identical to 0.035_01 release

0.035_01 2017-02-15 04:41:31 -0500
  - Add support for Alien::Build (plicease #181)

0.034 2017-02-15 04:30:35 -0500
  - Fixed regression introduced in 0.032 which broke repos as array
    reference

0.032 2017-02-09 13:36:07 -0500
  - Automatically add SSL modules as prereqs if we can tell with 
    certainity that we will need to use HTTP::Tiny with SSL (plicease 
    gh#178)
  - Add environment override of repository settings (gh#180 nshp++)

0.031 2017-01-28 08:11:06 -0500
  - Production release identical to 0.030_02 release

0.030_02 2017-01-25 06:25:26 -0500
  - Added a version method to Alien::Base.
  - Improved diagnostic when no files can be found on an attempted share install
  - A number of documentation improvements, as well as several new FAQs
  - Tests were failing on OpenBSD to do idiosyncratic pkg-config on that
    platform.  There is now a work around in the test for these idiosyncrasies.

0.030_01 Wed Nov 23, 2016
  - Fix test suite to handle it when . is not in @INC

0.030  Mon Oct 31, 2016
  - Production release identical to 0.029_01 release

0.029_01  Fri Oct 21, 2016
  - Fixed example in documentation for ExtUtil::MakeMaker
  - Fixed bug in test suite with where multiple users installing
    AB could cause failures
  - Added interface for overriding config.site variables
    see alien_extra_site_config

0.029  Thu Jun 16, 2016
  - Production release identical to 0.028_01 release
  - This is the "Happy Captain Picard Day" Release.

0.028_01  Tue Jun 14, 2016
  - Fixed bug where a colon (:) and equal sign (=) can confuse pkg-config
    parser (dakkar++ gh#165)

0.028  Mon Mar 14, 2016
  - Production release identical to 0.027_03 release

0.027_03  Wed Mar 9, 2016
  - Updated documentation to recommend the use of Alien::Base::ModuleBuild
    as a configure_requires.  PLEASE UPDATE YOUR Build.PL FILES!
  - Fixed bug where missing trailing slash when used with exact_filename
    would cause a failure (plicease gh#161, gh162)
  - Documentation fix in FAQ (plicease gh#159,gh#160)
  - Added compatability for Alien::Builder (plicease gh#158, see also gh#157)

0.027_02  Sat Feb 27, 2016
  - Fix bug where default ffi_name was incorrectly computed

0.027_01  Fri Feb 19, 2016
  - Deprecate %p
  - Require Alien::CMake 0.07 as a minimum when used as alien_bin_requires
    for Alien::Base compatability

0.027  Thu Feb 4, 2016
  - Production release identical to 0.026_02 release

0.026_02  Mon Feb 1, 2016
  - Fix test bug introduced in 0.026_01 where t/builder.t
    would fail on Windows if Alien::MSYS was not installed.

0.026_01  Mon Feb 1, 2016
  - Added alien_env property to Alien::Base::ModuleBuild
  - require HTTP::Tiny 0.044 for correct calculation of
    relative URLs on may websites.

0.026  Fri Jan 22, 2016
  - For http, use base URL from index request when downloading
    files (plicease gh#150, gh#151)

0.025  Wed Jan 20, 2016
  - Production release identical to 0.024_02 release

0.024_02  Mon Jan 18, 2016
  - Silenced warnings that can happen when multiple .pc files are included
    with a package (salva++ gh#148)
  - Fixed bug where verbose diagnostic could cause false positives
    and false negatives for system libraries (salva++ gh#147, plicease gh#149)

0.024_01  Tue Jan 12, 2016
  - Use URI internally for improved support for GitHub as a source
    (among others) (salva++ gh#144)

0.024  Thu Jan 7, 2015
  - Production release identical to 0.023_01 release

0.023_01  Sat Jan 2, 2015
  - Fixed a usually harmless but frequently annoying isssue where the
    alien_install_commands were executed on every invocation of './Build',
    './Build test', and './Build install' instead of just once as is
    needed.  (plicease gh#141)
  - Archive extraction can now be overridden with the alien_extract_archive
    method in Alien::Base::ModuleBuild. (salva++ gh#142)
  - Fixed bug with windows where using \ instead of / broke relocatable
    installs.  (plicease gh#139)
  - Promoted _env_do_system a public method named alien_do_system
    (salva++ gh#143)

0.023  Mon Sep 14, 2015
  - Fixed typo in FAQ
  - Updated FAQ Alien::gmake example to require Alien::gmake 0.11
  - Otherwise a production release identical to the 0.22_01 release

0.022_01  Fri Jul 31, 2015
  - Add support for ALIEN_INSTALL_TYPE environment variable

0.022  Mon Jul 20, 2015
  - Correction for the section "How do I specify a minumum or exact version
    requirement for packages that use pkg-config?" in the FAQ.

0.021_01  Wed Jul 15, 2015
  - Added a default %{pkg_config} helper
  - Fixed bug introduced in 0.016_01 where using --destdir or $ENV{DESTDIR}
    would break the "./Build install" command.

0.021  Wed Jul 15, 2015
  - Fixed bug where upgrading to 0.020 could break Alien modules installed

Makefile.PL
use strict;
use warnings;
use Config;
use ExtUtils::MakeMaker;

my %make_args = (
  NAME             => 'Alien::Base',
  DISTNAME         => 'Alien-Base',
  LICENSE          => 'perl',
  MIN_PERL_VERSION => '5.008001',
  VERSION_FROM     => 'lib/Alien/Base.pm',
  AUTHOR           => [
    'Graham Ollis <plicease@cpan.org>',
    'Joel A. Berger <joel.a.berger@gmail.com>',
  ],

  CONFIGURE_REQUIRES => {
    'ExtUtils::MakeMaker' => 0,
  },
  PREREQ_PM => {
    'Capture::Tiny'             => 0.17,
    'File::ShareDir'            => 0,
    'Text::ParseWords'          => '3.26',
    'FFI::CheckLib'             => 0.11,
    'JSON::PP'                  => 0,
    'Alien::Base::ModuleBuild'  => '0.040',
  },
  BUILD_REQUIRES => {},
  TEST_REQUIRES => {
    'Test2::V0'                 => '0.000071',
    'Test2::Suite'              => '0.000071',
    'Test2'                     => '1.302015',
    'File::chdir'               => 0.1005,
  },

  META_MERGE => {
    'meta-spec' => { version => 2 },
    resources => {
      repository => {
        type => 'git',
        url  => "http://github.com/Perl5-Alien/Alien-Base.git",
        web  => "http://github.com/Perl5-Alien/Alien-Base",
      },
      bugtracker => {
        web => "http://github.com/Perl5-Alien/Alien-Base/issues",
      },
      homepage => "https://metacpan.org/pod/Alien::Base",
      x_MailingList => "https://groups.google.com/forum/#!forum/perl5-alien",
      x_IRC => "irc://irc.perl.org/#native",
    },
    no_index => {
      directory => [
        "corpus",
        "maint",
      ],
    },
    dynamic_config => 1,
  },

  clean => {
    FILES => '_Inline Alien-Base-*',
  },
);

if($ENV{ALIEN_BASE_NO_ABMB})
{
  delete $make_args{'PREREQ_PM'}->{'Alien::Base::ModuleBuild'};
}

unless(`pkg-config --version` && $? == 0) {
  $make_args{'PREREQ_PM'}->{'PkgConfig'} = '0.07520';
}

# For now we prefer PkgConfig.pm over pkg-config on
# Solaris 64 bit Perls.  We may need to do this on
# other platforms, in which case this logic should
# be abstracted so that it can be shared here and
# in lib/Alien/Base.pm#pkg_config_command
if($^O eq 'solaris' && $Config{ptrsize} == 8) {
  $make_args{'PREREQ_PM'}->{'PkgConfig'} = '0.08826';
}


my %fallback = (
  %{ $make_args{TEST_REQUIRES} },
  %{ $make_args{BUILD_REQUIRES} },
  %{ $make_args{PREREQ_PM} },
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  $make_args{PREREQ_PM} = \%fallback;
  delete $make_args{TEST_REQUIRES};
  delete $make_args{BUILD_REQUIRES};
}

delete $make_args{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%make_args);
